# Python Machine Learning 7
# Grid Search
## *By Sal Lascano*

Okay! lets say that we are working on a project. We used cross validation to determine which model we will be using on our data. According to cross validation the most suitable model will be Lasso. Now its time to find out the best value for our parameter/lambda. We could try all values manually to see which one is the best but that will take a very long time. This is were we use Grid Search, a technic that helps us find the appropriate value of our Lambda.

Grid Search tests every single possible parameter we ask it to and validates it via cross validation to see out of the ones it has tested which one is the best parameter for the specific problem. We learned about cross validation on notebook 4.

Grid Search can be used also on the KNN algorithm (which we will talk about on notebook 10) to find out the appropriate number for K and to see its score.

Lets walk through a example.

In [9]:
#lets strat by importing the necesary libraries
from sklearn import neighbors
from sklearn.model_selection import GridSearchCV
from sklearn import datasets
knn = neighbors.KNeighborsClassifier()

In [3]:
#get that famous iris data
iris = datasets.load_iris()

# set the possible parameters or K from 3 to 30
grid_param = [{'n_neighbors': range(3, 31)}]

# fit all models, we will be using 5 folds on the cross validation as cv=5.
# selecting accuracy for the scoring to get the best possible parameter
para_search = GridSearchCV(estimator=knn, param_grid=grid_param, scoring='accuracy', cv=5, return_train_score=True)
para_search = para_search.fit(iris.data, iris.target)

In [4]:
# cv_results_ stores a lot of information about the cross validation, lets print it
print(list(para_search.cv_results_.keys()))
para_search.cv_results_

['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_n_neighbors', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'split3_test_score', 'split4_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score', 'split0_train_score', 'split1_train_score', 'split2_train_score', 'split3_train_score', 'split4_train_score', 'mean_train_score', 'std_train_score']


{'mean_fit_time': array([0.00045223, 0.00026994, 0.0003705 , 0.00035   , 0.00027766,
        0.00023022, 0.00023456, 0.00022202, 0.00024409, 0.00024014,
        0.00021844, 0.00021405, 0.00022235, 0.00020785, 0.000212  ,
        0.00021319, 0.00020914, 0.00021582, 0.00022097, 0.00021572,
        0.00020642, 0.00022202, 0.0001977 , 0.00020528, 0.00029216,
        0.00022635, 0.00021052, 0.00019817]),
 'mean_score_time': array([0.00068336, 0.00050206, 0.00067964, 0.00061793, 0.00048614,
        0.00044279, 0.00044508, 0.00045681, 0.00044322, 0.00048299,
        0.00045557, 0.00045614, 0.00046129, 0.0004427 , 0.00044723,
        0.00045185, 0.0004683 , 0.00045719, 0.00047822, 0.00046978,
        0.00043797, 0.00044627, 0.00044103, 0.00044789, 0.0006454 ,
        0.00049052, 0.00043759, 0.00043244]),
 'mean_test_score': array([0.96666667, 0.97333333, 0.97333333, 0.98      , 0.98      ,
        0.96666667, 0.97333333, 0.98      , 0.98      , 0.98      ,
        0.97333333, 0.96666667, 0.966

In [5]:
# lets check out the scores of the parametes
para_search.grid_scores_

#Check out how k=7 & 7 have the same scores but given that we asked for more decimals we probably
#would be able to see that perhaps after many decimals, k=6 is better than k=7

/Users/sals/anaconda3/lib/python3.6/site-packages/sklearn/model_selection/_search.py:761: DeprecationWarning: The grid_scores_ attribute was deprecated in version 0.18 in favor of the more elaborate cv_results_ attribute. The grid_scores_ attribute will not be available from 0.20
  DeprecationWarning)


[mean: 0.96667, std: 0.02108, params: {'n_neighbors': 3},
 mean: 0.97333, std: 0.01333, params: {'n_neighbors': 4},
 mean: 0.97333, std: 0.02494, params: {'n_neighbors': 5},
 mean: 0.98000, std: 0.01633, params: {'n_neighbors': 6},
 mean: 0.98000, std: 0.01633, params: {'n_neighbors': 7},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 8},
 mean: 0.97333, std: 0.02494, params: {'n_neighbors': 9},
 mean: 0.98000, std: 0.02667, params: {'n_neighbors': 10},
 mean: 0.98000, std: 0.02667, params: {'n_neighbors': 11},
 mean: 0.98000, std: 0.02667, params: {'n_neighbors': 12},
 mean: 0.97333, std: 0.02494, params: {'n_neighbors': 13},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 14},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 15},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 16},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 17},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 18},
 mean: 0.96667, std: 0.02981, params: {'n_neighbors': 19},
 mea

In [6]:
#lets print the set score
print(para_search.best_score_)

#and wich K gave us that score
print(para_search.best_params_)

0.98
{'n_neighbors': 6}


The best parameter is 6 corresponding to an accuracy of 98%